tensorflow/compiler/mlir/xla/transforms/cycle_detector.cc

```diff
@@ -0,0 +1,335 @@
+  bool visited;  // Temporary marker used by depth-first-search
+  void* data;    // User-supplied data
+  OrderedNodeSet in;   // List of immediate predecessor nodes in graph
+  OrderedNodeSet out;  // List of immediate successor nodes in graph
+};
+
+}  // namespace
+
+
+struct GraphCycles::Rep {
+  Vec<Node*> nodes_;
+  Vec<int32_t> free_nodes_;  // Indices for unused entries in nodes_
+
+  // Temporary state.
+  Vec<int32_t> deltaf_;  // Results of forward DFS
+  Vec<int32_t> deltab_;  // Results of backward DFS
+  Vec<int32_t> list_;    // All nodes to reprocess
+  Vec<int32_t> merged_;  // Rank values to assign to list_ entries
+  Vec<int32_t> stack_;   // Emulates recursion stack when doing depth first search
+};
+
+GraphCycles::GraphCycles() : rep_(new Rep) {}
+
+GraphCycles::~GraphCycles() {
+  for (Vec<Node*>::size_type i = 0; i < rep_->nodes_.size(); i++) {
+    delete rep_->nodes_[i];
+  }
+  delete rep_;
+}
+
+int32_t GraphCycles::NewNode() {
+  if (rep_->free_nodes_.empty()) {
+    Node* n = new Node;
+    n->visited = false;
+    n->data = nullptr;
+    n->rank = rep_->nodes_.size();
+    rep_->nodes_.push_back(n);
+    return n->rank;
+  } else {
+    // Preserve preceding rank since the set of ranks in use must be
+    // a permutation of [0,rep_->nodes_.size()-1].
+    int32_t r = rep_->free_nodes_.back();
+    rep_->nodes_[r]->data = nullptr;
+    rep_->free_nodes_.pop_back();
+    return r;
+  }
+}
+
+void GraphCycles::RemoveNode(int32_t node) {
+  Node* x = rep_->nodes_[node];
+  for (int32_t y : x->out.GetSequence()) {
+    rep_->nodes_[y]->in.Erase(node);
+  }
+  for (int32_t y : x->in.GetSequence()) {
+    rep_->nodes_[y]->out.Erase(node);
+  }
+  x->in.Clear();
+  x->out.Clear();
+  rep_->free_nodes_.push_back(node);
+}
+
+bool GraphCycles::HasEdge(int32_t x, int32_t y) const {
+  return rep_->nodes_[x]->out.Contains(y);
+}
+
+void GraphCycles::RemoveEdge(int32_t x, int32_t y) {
+  rep_->nodes_[x]->out.Erase(y);
+  rep_->nodes_[y]->in.Erase(x);
+  // No need to update the rank assignment since a previous valid
+  // rank assignment remains valid after an edge deletion.
+}
+
+static bool ForwardDFS(GraphCycles::Rep* r, int32_t n, int32_t upper_bound);
+static void BackwardDFS(GraphCycles::Rep* r, int32_t n, int32_t lower_bound);
+static void Reorder(GraphCycles::Rep* r);
+static void Sort(const Vec<Node*>&, Vec<int32_t>* delta);
+static void MoveToList(GraphCycles::Rep* r, Vec<int32_t>* src, Vec<int32_t>* dst);
+static void ClearVisitedBits(GraphCycles::Rep* r, const Vec<int32_t>& nodes);
+
+bool GraphCycles::InsertEdge(int32_t x, int32_t y) {
+  if (x == y) return false;
+  Rep* r = rep_;
+  Node* nx = r->nodes_[x];
+  if (!nx->out.Insert(y)) {
+    // Edge already exists.
+    return true;
+  }
+
+  Node* ny = r->nodes_[y];
+  ny->in.Insert(x);
+
+  if (nx->rank <= ny->rank) {
+    // New edge is consistent with existing rank assignment.
+    return true;
+  }
```

Comment:
We always seem to hit the return true above and never fallback to the code below.

Reply:
Actually, it will. At GraphCycles construction stage below branch will not be called, while in edge contract stage we may insert  new edges and the rank may need to recompute accordingly.

For example, for a graph (0 -> 1, 0 -> 2, here the number is the rank)， once we contract edge 0 -> 2 and choose node `2` as the representation of the fused pattern, then we need to insert an edge (2 -> 1)，this triggers rank re-computation and all these relevant functions.

Reply:
Sorry if I wasn't clear: I'm not saying this is dead code, I was just pointing out that the *current fusion tests* don't have coverage for the kind of case you describe.
The question was if you can expand the fusion cases to cover more of this?